### PR TITLE
Malyan M200 v1 / v2 working builds

### PR DESCRIPTION
**Pull request by thinkyhead**
This PR is a work in progress to get working Malyan M200 builds for both V1 and V2 boards.

### CODE REVIEW

**Comment by xC0000005** [comment]
@mojocorp, @ppetermann, relevant to your efforts to build M300 automatically.

**Comment by mojocorp** [comment]
@thinkyhead I can rework this PR if you want. The M200v2 hardware is really close to my M300.

I have a working M200/M200v2 autobuild: https://github.com/papabricole/Marlin/tree/malyan-m200-autobuild

**Comment by 3Dangel** [comment]
> @thinkyhead I can rework this PR if you want. The M200v2 hardware is really close to my M300.
> 
> I have a working M200/M200v2 autobuild: https://github.com/papabricole/Marlin/tree/malyan-m200-autobuild
> 
> see commit: [papabricole@2f1ecda](https://github.com/papabricole/Marlin/commit/2f1ecda700589c6f952412534614b62c2a0c3f79)

hello @mojocorp 
I'm wondering how the autobuild at your branch works?
I just checked, and all the settings are set as default for atmega2560...
would you be willing to guide me a little be how to get the Mini Select V2_Marlin to compile?
so I need to manually change platformio's env? and copy the config Examples for Malyan 200 V2 from Marlin configuration folder @ github?

Thanks for your help!!
Cheers

**Comment by mojocorp** [comment]
@3Dangel there should be plenty of tutorials how to build marlin with platformio and i cannot do better in just a comment. If you cherry-picked my commit and changed the config files then do: platformio run -e malyan_M200v2

**Comment by 3Dangel** [comment]
Thanks for your quick reply.
:-)
I know how to use atom/platformio and Vscode.
 Just needed to know if yours is something different.
Because I tried and it did not compile.
- changing platformio.ini
-bringing the example config file from Marlin
Lots of yellow warning and it did not want to compile.

So you said you have it working for your v2.
We’re you using your repo and compiled successfully?
Thanks a lot for sharing and spending your time on this!!

**Comment by mojocorp** [comment]
I you are on a fresh marlin bugfix-2.0.x , cherry-pick my single commit and change the Configuration.h with the one for the M200v2 then it should just compile.
I don't do anything special. If you have an error like 
"#error "SERIAL_PORT cannot be 0. (Port 0 does not exist.) Please update your configuration."" 
Then it means you put the wrong Configuration.h

File: buildroot/share/PlatformIO/boards/malyanM200v2.json [hunk 1]
```diff
@@ -1,10 +1,11 @@
 {
   "build": {
     "cpu": "cortex-m0",
-    "extra_flags": "-DSTM32F070xB",
+    "extra_flags": "-DNO_UNUSED -DSTM32F070xB",
     "f_cpu": "48000000L",
     "mcu": "stm32f070rbt6",
-    "genericvariant" : "MALYAN_M200_V2",
+    "genericvariant" : "MALYANM200_F070CB",
```
Comment (xC0000005) [inline buildroot/share/PlatformIO/boards/malyanM200v2.json]: "genericvariant" shouldn't mean anything in this Arduino core - it's left over from me adding PIO support for STM32GENERIC, last time I checked (unless someone has changed the builder).

Comment (mojocorp) [reply buildroot/share/PlatformIO/boards/malyanM200v2.json]: I removed the file entirely in my commit, its not necessary. It's just a tiny problem in platformio.ini, nothing more.

**Comment by thinkyhead** [comment]
So, I've been working on other PRs and just getting back to this one. Please feel free to add on to it. I think GitHub now has a "Suggest Change" option in the top-right dohickey while viewing changes under the Files tab. You make an edit and it shows up inline as a suggestion. When accepted it gets added as a new commit.

**Comment by mojocorp** [comment]
@thinkyhead may I propose a minimal M200 fix PR instead ? There are other c++ changes in this PR not necessary to make it build, or I missed something. I can craft the PR to minimize conflicts with this PR.

**Comment by jValdron** [comment]
> @thinkyhead I can rework this PR if you want. The M200v2 hardware is really close to my M300.
> 
> I have a working M200/M200v2 autobuild: https://github.com/papabricole/Marlin/tree/malyan-m200-autobuild

Using your platformio.ini's env for `malyan_M200v2` allowed me to build the latest `bugfix-2.0.x` for my M200 v2, however, the firmware.bin ended up being only 20kb (typically around 100kb using Arduino IDE).

I'll dig in when I get a chance and try to come up with a working configuration for my board/setup.

EDIT: Adding output:
```
Building .pio\build\malyanM200v2\firmware.bin
Checking size .pio\build\malyanM200v2\firmware.elf
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [===       ]  31.8% (used 4820 bytes from 15168 bytes)
Flash: [==        ]  15.7% (used 19308 bytes from 122880 bytes)
```

**Comment by mojocorp** [comment]
> > @thinkyhead I can rework this PR if you want. The M200v2 hardware is really close to my M300.
> > I have a working M200/M200v2 autobuild: https://github.com/papabricole/Marlin/tree/malyan-m200-autobuild
> 
> Using your platformio.ini's env for `malyan_M200v2` allowed me to build the latest `bugfix-2.0.x` for my M200 v2, however, the firmware.bin ended up being only 20kb (typically around 100kb using Arduino IDE).
> 
> I'll dig in when I get a chance and try to come up with a working configuration for my board/setup.
> 
> EDIT: Adding output:
> 
> ```
> Building .pio\build\malyanM200v2\firmware.bin
> Checking size .pio\build\malyanM200v2\firmware.elf
> Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
> RAM:   [===       ]  31.8% (used 4820 bytes from 15168 bytes)
> Flash: [==        ]  15.7% (used 19308 bytes from 122880 bytes)
> ```

Strange, I just tried and everything seems fine:

RAM:   [=====     ]  53.9% (used 8168 bytes from 15168 bytes)
Flash: [========= ]  85.6% (used 105128 bytes from 122880 bytes)

**Comment by jValdron** [comment]
@mojocorp, thanks for confirming. Just tried on a fresh VS Code install in a Linux VM, built fine this time:
```
Building .pio/build/malyan_M200v2/firmware.bin
Checking size .pio/build/malyan_M200v2/firmware.elf
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [======    ]  56.4% (used 8556 bytes from 15168 bytes)
Flash: [========= ]  94.6% (used 116296 bytes from 122880 bytes)
```

Compared to my existing VS Code + platform.io on my Windows machine, that I had tried to get working months back. I've never had much luck building Marlin on Windows.

Tried to flash it and it's currently not working, but I haven't ran bugfixes-2.0.x yet on my machine, I had been using the m200fixes branch from xC0000005.

**Comment by mojocorp** [comment]
> @mojocorp, thanks for confirming. Just tried on a fresh VS Code install in a Linux VM, built fine this time:
> 
> ```
> Building .pio/build/malyan_M200v2/firmware.bin
> Checking size .pio/build/malyan_M200v2/firmware.elf
> Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
> RAM:   [======    ]  56.4% (used 8556 bytes from 15168 bytes)
> Flash: [========= ]  94.6% (used 116296 bytes from 122880 bytes)
> ```
> 
> Compared to my existing VS Code + platform.io on my Windows machine, that I had tried to get working months back. I've never had much luck building Marlin on Windows.
> 
> Tried to flash it and it's currently not working, but I haven't ran bugfixes-2.0.x yet on my machine, I had been using the m200fixes branch from xC0000005.

Could you be more precise on what is not working ? Not working at all ? crashing ?

**Comment by jValdron** [comment]
> Could you be more precise on what is not working ? Not working at all ? crashing ?

Hey @mojocorp, sorry for being sparse on details, figured this was not a problem related to this PR. But if you guys are willing to help, that'd be great! Gist of it is that it crashes on boot.

So, last build I had on my printer, which worked properly, seems to be e03432b03edd3064f8690157d871cdf0d3b5ae05 from xC0000005, seems to be the latest commit from m200fixes.

I tried building the latest bugfixes-2.0.x yesterday, but it had an issue with the ExtUI/HAS_MESH (which you guys resolved super quickly #17834!), I just added the condition and tested it out myself, but it crashed on boot. My [UI controller started](https://imgur.com/3y8qm62), as usual, then when Marlin "kicks in", it just goes to [this screen](https://imgur.com/QSLWrHF) and loops back to home screen, then back to that "finish" screen.

So it just seems like Marlin crashes, tries to start back, crashes, etc.

Just tried the latest bugfixes-2.0.x today and it seems to do the same thing. I've also tried using the stock M200 Configuration/Configuration_adv files (I have a somewhat modded MPSM v2), and no go, same thing. Also tried with nothing plugged into my controller (except the LCD), and same thing.

Here's my configuration files (although it doesn't seem like that's the issue): https://pastebin.com/1FZ6UPCW / https://pastebin.com/E5EA6Vrf

I use OctoPrint as a host, and here's the full logs of when I connect to it (I can connect to it for about 2-3s): https://pastebin.com/TKDkngYu

Snippet of before it crashes:
```
Recv: echo:enqueueing "S:I"
Recv: echo:Unknown command: "S:I"
Recv:  T:0.00 /0.00 B:0.00 /0.00 @:0 B@:0
Recv: echo:enqueueing "S:I"
Recv: echo:Unknown command: "S:I"

Unexpected error while reading serial port, please consult octoprint.log for details: SerialException: 'device reports readiness to read but returned no data (device disconnected or multiple access on port?)' @ comm.py:_readline:2793
Please see https://faq.octoprint.org/serialerror for possible reasons of this.
Changing monitoring state from "Operational" to "Offline (Error: SerialException: 'device reports readiness to read but returned no data (device disconnected or multiple access on port?)' @ comm.py:_readline:2793)"
Connection closed, closing down monitor
```

So seems to send temperature status, but it also seems to be receiving UI commands for some reason. However, this could be an OctoPrint quirk and benign error, I don't remember if those were there before or not.

Sorry for the very detailed post :) Don't have much experience debugging Marlin/controllers in general, so if you could point me in the right direction or whatnot, that'd be great!

**Comment by mojocorp** [comment]
@jValdron you are right, it's not working anymore.
I just tried the current bugfix-2.0.x and my M300 stopped working. I'm currently based onto
50fa87f1cfad92ff88f7c14ecaa39e1c2b33639f and my M300 work fine.
Since the M300 and M200v2 harware is very close, you may have the same problem.
I rebased my branch malyan-m200-autobuild onto a know working Marlin. Give it a try.

**Comment by jValdron** [comment]
Just checked out your `malyan-m200-autobuild` branch, dumped my configurations in there, and no dice for me. Seems to be crashing too, same kind of behavior, but I don't see the "finish" screen, just loops on the home screen.

Might revert back to `m200fixes` until I have more time (and expertise lol) to look into this.

Full log: https://pastebin.com/VLBas5fQ

Nothing really stands out from the logs.

**Comment by xC0000005** [comment]
A PR that was meant to provide default timer overrides may have broken this.

**Comment by mojocorp** [comment]
@xC0000005 It's what I though when looking at the history but it was not the problem. It was just a tiny bug in Malyan ExtUI LCD. I've already created a PR.

**Comment by xC0000005** [comment]
I'm getting close to getting auto-timer working. Still need to fix IRQ
numbers.

https://github.com/xC0000005/Marlin/tree/AutoSelectTimers

**Comment by thinkyhead** [comment]
I've just rebased this on the latest code which has the LCD patch. There's also a patch for the M200 in the bugfix branch ahead of this PR, so you might have luck with bugfix just as it is now. Please give both a try to see whether we're getting any closer.

**Comment by jValdron** [comment]
I'm thinking that @xC0000005 is right and it's something with the default timer overrides that broke things on my M200 v2.

Tried to build the latest `bugfix-2.0.x` and the `thinkyhead:bf2_malyan_v2_PR` branch with both Arduino IDE and Platform.io (never had a working build on my printer with Platform.io thus far). Still doesn't seem to work. Here's the Octoprint logs, not much available in there, as usual: https://pastebin.com/kjdbZXqq

I tried with my old firmware (from `m200fixes`) back just to make sure it was fine, and it flashed it/booted right up.

FYI, still need the Platform.io environment `env:malyan_M200v2` from https://github.com/papabricole/Marlin/blob/malyan-m200-autobuild/platformio.ini to build within VS Code.

**Comment by thinkyhead** [comment]
Timers defined in the pins file are:
```cpp
#define STEP_TIMER 1
#define TEMP_TIMER 3
```

**Comment by thinkyhead** [comment]
> I just tried the current bugfix-2.0.x and my M300 stopped working. I'm currently based onto
50fa87f and my M300 work fine.

@mojocorp — At that point in time the timer overrides in pins files were not working, so instead of timers 6 and 7 your build would be using timers 16 and 17.

**Comment by JamesHagerman** [comment]
What is the status of this work? Can I help test anything?

I've been trying to build Marlin for my Malyan M200 (MPSM v2) and I'm not sure which set of PR/branches, commits, and config files I should be using to get things to work.

I've tried compiling the latest `bugfix-2.0.x` with the latest M200 configs from `Marlin/Configurations`, I've tried with PlatformIO, I've tried with the Arduino IDE, and I've tried on Windows and Linux. Some of my attempts have compiled successfully, but then the firmware doesn't behave as expected.

**Comment by jValdron** [comment]
Haven't been able to get anything newer working than the `m200fixes` branch from @xC0000005: https://github.com/xC0000005/Marlin/tree/M200Fixes

I've been using Arduino 1.8 and an older version of the STM board library.

Everything else, that I have been able to compile, either just crashes shortly after booting, or doesn't boot at all.

**Comment by xC0000005** [comment]
Ok, got it compiling, next I'll see if it actually works. I'm looking at the 070 board first (M200).

**Comment by xC0000005** [comment]
@mojocorp, what was the watchdog bug you said you'd seen? I have a build that boots and then resets, and I'm wondering if it's the same problem.

**Comment by xC0000005** [comment]
So, the reason this code doesn't work (yet) is that the interrupts aren't running, ever (no idea why, yet). So the printer comes up, the watchdog is on, the temp ISR never executes, and the printer reboots (GOTO 10)

Once I figure that out, we should be ok, I have both 103 and 070 building/running to the same issue.

**Comment by sjasonsmith** [comment]
@xC0000005, what version of the ststm32arduino framework is it using? If it is using 1.8 (4.108000), there are some regressions in the timer code that require some changes to work.

**Comment by xC0000005** [comment]
I did not set a version. Is there a proper version to clamp to?

**Comment by sjasonsmith** [comment]
Are you building these in PlatformIO or Arduino? In Arduino you just select 1.7 of the framework. In platformio everything except the M300 should be locked to 4.10700 (1.7) in the current bugfix-2.0.x. If you aren't in sync with that right now you can force it in a single environment by adding the following:

`platform_packages = framework-arduinoststm32@>=4.10700,<4.10800`

If you want/need to use 4.10800 (1.8), you can look at the timer changes I made in the following PR, which got them working on a SKR Pro.
https://github.com/MarlinFirmware/Marlin/pull/17970

That PR isn't going in because that version of the framework seems to reduce I2C EEPROM reliability. These timer changes are also for the most part regressions which weren't necessary in 1.7, and will not necessary in 1.9.

**Comment by xC0000005** [comment]
These produce functional builds (though I'm still having issues with the timer firing too fast to move the steppers:
```ini
#
# Malyan M200 (STM32F103CB)
#
[env:STM32F103CB_malyan]
platform          = ${common_stm32.platform}
extends     = common_stm32
board       = malyanm200_f103cb
build_flags = ${common_stm32.build_flags}
  -DUSBCON  -DUSBD_PID=0x2EAF -DUSBD_VID=0x0483 -DUSB_MANUFACTURER=\"Unknown\" -DUSB_PRODUCT=\"MALYAN_M200\"
  -DHAL_PCD_MODULE_ENABLED -DUSBD_USE_CDC -DDISABLE_GENERIC_SERIALUSB -DHAL_UART_MODULE_ENABLED
  -DCUSTOM_STARTUP_FILE -DBLAH
lib_ignore =   SoftwareSerialM
src_filter  = ${common.default_src_filter} +<src/HAL/STM32>

#
# Malyan M200 v2 (STM32F070RB)
#
[env:STM32F070RB_malyan]
platform          = ${common_stm32.platform}
extends     = common_stm32
board       = malyanm200_f070cb
build_flags = ${common_stm32.build_flags}
  -DUSBCON  -DUSBD_PID=0x1EAF -DUSBD_VID=0x0483 -DUSB_MANUFACTURER=\"Unknown\" -DUSB_PRODUCT=\"MALYAN_M200\"
  -DHAL_PCD_MODULE_ENABLED -DUSBD_USE_CDC -DDISABLE_GENERIC_SERIALUSB -DHAL_UART_MODULE_ENABLED
  -DCUSTOM_STARTUP_FILE -DBLAH
lib_ignore =   SoftwareSerialM
src_filter  = ${common.default_src_filter} +<src/HAL/STM32>
```

**Comment by thinkyhead** [comment]
When there are definitive answers, this might get fixed up. But until that time, this PR is not trusted and will not be merged.

**Comment by sjasonsmith** [comment]
I'm thinking that after 2.0.6 releases, we could merge in my changes to use the newer STM32 framework. It is possible that not every board will work perfectly, but we can than fix them up in smaller focused changes.

What do you think of that approach @xC0000005, since you seem to work with the boards the most likely to have issues (Lerdge, Malyan).

**Comment by xC0000005** [comment]
This seems reasonable. I will be away for a week with no computer access, so I will see where we are at when I return.

**Comment by nick-diel** [comment]
I have a M200 V2 and am happy to help with testing.  I had been compiling Marlin with Arduino CLI and STM32duino 1.7 (also worked with the Master branch from 3 months ago).  Everything functioned, just was dealing with layer shifting (same gcode had no layer shifting with OEM firmware).  Spent a bunch of time slowing down acceleration, stepper pulses, s-curve, junction deviation, etc. and never really solved my problem.  Put the printer on pause for a few months and I am now hoping some of the recent changes fix the layer shifting.

@sjasonsmith is the best use of my time to test #18496?  Do I need to specify a specific version of STM32duino to avoid the watchdog bug?

**Comment by sjasonsmith** [comment]
@nick-diel I’m not sure if @xC0000005  comment about the timer issue is with the latest state of that branch or not. It would be great to figure out what is preventing them from working quite right. I’m not sure if there is anything in this branch that needs merged into it. Testing my PR would be great, but it’s possible that @xC0000005 will already know the expected outcome.

**Pull request closed without merging**